website/content/docs/job-specification/numa.mdx

```diff
@@ -0,0 +1,76 @@
+---
+layout: docs
+page_title: numa Block - Job Specification
+description: |-
+  The "numa" block is used configure NUMA aware scheduling strategy for a task.
+---
+
+# `numa` Block
+
+<Placement groups={['job', 'group', 'task', 'resources', 'numa']} />
+
+The `numa` block is used to configure how Nomad will assign CPU cores for a task
+while taking the [NUMA hardware topology][numa_wiki] of a node into consideration.
+Workloads that are sensitive to memory latency can perform significantly better
+when pinned to CPU cores on the same NUMA node.
+
+<EnterpriseAlert>
+
+This functionality only exists in Nomad Enterprise. This is not
+present in the source available version of Nomad.
+</EnterpriseAlert>
+
+<Note>
+
+NUMA aware scheduling is currently limited to Linux.
+</Note>
+
+```hcl
+job "example" {
+  group "group" {
+    task "task" {
+      resources {
+        cores = 8
+        numa {
+          affinity = "require"
+        }
+      }
+    }
+  }
+}
+```
+
+In the example above, the task is requesting Nomad to reserve 8 CPU cores, all
+of which must be colocated on the same hardware NUMA node. The Nomad scheduler
+will ensure the task is placed on a Nomad node with sufficient unused CPU cores
+in a compatible configuration.
+
+Configuring the `numa` block requires the task specifies CPU resources using
+the [`cores`][cores] parameter.
+
+# `numa` Parameters
+
+- `affinity` `(string: "none")` - Specifies the strategy Nomad will use when
+  selecting CPU cores to assign to a task. Possible values are `"none"`,
+  `"prefer"`, or `"require"`.
+  - `none` - Nomad is free to allocate CPU cores using any strategy. Nomad uses
+  this freedom to allocate cores in such a way that minimizes the amount of
+  fragmentation of core availability per NUMA node.
+  - `prefer` - Nomad will select the set of CPU cores on a node that minimizes
+  the total distance between those cores, but does not limit those CPU core
+  selections to come from a single NUMA node.
+  - `require` - Nomad will select a set of CPU cores that are strictly colocated
+  on the same hardware NUMA node. If there are no Nomad nodes with a sufficient
+  number of available cores in a compatible configuration, task placement will
+  fail due to exhausted resources.
+
+<Note>
+
+The `require` affinity option should be used sparingly due to
+the implied fragmentation caused by reserving CPU cores based on the NUMA node
+they are associated with. Use it for workloads known to be highly sensitive
+to memory latencies.
```

Comment:
> ...used sparingly... ...implied fragmentation...

I know people often want more prescriptive guidance from us, but I would rather frame this in terms of the tradeoff being presented than to discourage the use of a feature that could reduce overall resource consumption dramatically. (Assuming if you're avoiding 300% performance penalties due to cross-node latencies, you can run fewer instances of a service to serve the same number of requests.)

So perhaps something like:

> The `require` affinity option may cause workload fragmentation by reserving CPU cores based on the NUMA node they are associated with. Use it for workloads known to be highly sensitive to memory latencies.

Might even be worth defining `workload fragmentation` somewhere as something like:

> A jobspec constraint that prevents optimal binpacking of Clients. This can waste cluster resources by leaving some Client resources free but unusable. For example when `numa.affinity = "require"`, workloads cannot be scheduled on Clients which may have ample free compute resources unless those compute resources happen to be colocated on a single NUMA node.

idk where an appropriate place for that would be though.